### PR TITLE
fix: [CDS-72250]: fixing file references in command and shell script provision step

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
File references was not being shown for file store when files are getting referenced in these steps
Implemented with file ref for these steps to extract file reference and added support for handling file reference in json creator.

Tests Done:
1. Adding files in command step and shell script provisioner step and getting reference in file store

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)

### CODE REVIEW

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java [hunk 1]
```diff
@@ -7,17 +7,98 @@
...
+      ParameterField<List<String>> refValue = ref.getValue();
+      List<String> files =
+          (isNull(refValue) || isNull(refValue.getValue())) ? Collections.emptyList() : refValue.getValue();
+      files.stream().filter(EmptyPredicate::isNotEmpty).forEach(scopedFilePath -> {
+        if (!EngineExpressionEvaluator.hasExpressions(scopedFilePath)) {
+          try {
+            FileReference fileReference =
+                FileReference.of(scopedFilePath, accountIdentifier, orgIdentifier, projectIdentifier);
+            Optional<FileDTO> fileDTO;
+            fileDTO = fileStoreService.getByPath(fileReference.getAccountIdentifier(), fileReference.getOrgIdentifier(),
```
[the remark this answers is not in the document]
Comment (yogesh-chauhan) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java:84]: TEST

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/CommandStepInfo.java [hunk 2]
```diff
@@ -83,4 +87,19 @@ public SpecParameters getSpecParameters() {
...
+
+  @Override
+  public Map<String, ParameterField<List<String>>> extractFileRefs() {
+    Map<String, ParameterField<List<String>>> fileRefMap = new HashMap<>();
+    List<String> fileRefs = new ArrayList<>();
+    commandUnits.forEach(commandUnit -> {
+      if (commandUnit != null && commandUnit.getSpec() instanceof ScriptCommandUnitSpec) {
+        ScriptCommandUnitSpec scriptCommandUnitSpec = (ScriptCommandUnitSpec) commandUnit.getSpec();
+        List<String> fileRef = scriptCommandUnitSpec.getSource().getFileRefs();
+        fileRefs.addAll(fileRef);
```
[the remark this answers is not in the document]
Comment (yogesh-chauhan) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/CommandStepInfo.java:99]: commandUnits.identifier...

**Review by yogesh-chauhan — COMMENTED** [review]
Resolving some of the comments as discussed internally

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java [hunk 3]
```diff
@@ -7,17 +7,98 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
+import lombok.extern.slf4j.Slf4j;
 
 @OwnedBy(HarnessTeam.PIPELINE)
+@Slf4j
 public class CDPMSStepFilterJsonCreatorV2 extends GenericStepPMSFilterJsonCreatorV2 {
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java:44]: done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java [hunk 4]
```diff
@@ -7,17 +7,98 @@
...
+  @Override
+  public FilterCreationResponse handleNode(FilterCreationContext filterCreationContext, AbstractStepNode yamlField) {
+    FilterCreationResponse response = super.handleNode(filterCreationContext, yamlField);
+    List<EntityDetailProtoDTO> result = response.getReferredEntities();
+
+    // add file reference for step if step implements with file ref interface
+    if (WithFileRefs.class.isAssignableFrom(yamlField.getStepSpecType().getClass())) {
+      Map<String, ParameterField<List<String>>> fileRefs =
+          ((WithFileRefs) yamlField.getStepSpecType()).extractFileRefs();
+      addFileReferredEntities(filterCreationContext, result, fileRefs);
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java:61]: done

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/CommandStepInfo.java` line 104 at r4](https://reviewable.io/reviews/harness/harness-core/49938#-NZS6voT5-_agVkbmCpU:-NZS6voT5-_agVkbmCpV:b-co96cx) ([raw file](https://github.com/harness/harness-core/blob/7cce0824b219e7bc968234eb8136d1e1f36a3129/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/CommandStepInfo.java#L104)):*
> ```Java
>       });
>     }
>     fileRefMap.put("commandUnits.identifier.spec.source.spec.file", ParameterField.createValueField(fileRefs));
> ```

I meant this can be the commandUnitIdentifier
String.format("commandUnits.%s.spec.source.spec.file", commandUnit.identifier)

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 2 of 5 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 1 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)* status: 7 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained.

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 1 of 5 files at r1, 3 of 6 files at r5, 1 of 1 files at r6, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)* status: 10 file(s) to review. 1 of 2 approvals obtained.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java [hunk 5]
```diff
@@ -7,17 +7,101 @@
...
+                  + ref.getKey() + PATH_CONNECTOR + YAMLFieldNameConstants.FILES;
+              String fileScopedIdentifier =
+                  FileReference.getScopedFileIdentifier(fileReference.getScope(), fileDTO.get().getIdentifier());
+
+              result.add(FilterCreatorHelper.convertToEntityDetailProtoDTO(accountIdentifier, orgIdentifier,
+                  projectIdentifier, fullQualifiedDomainName, ParameterField.createValueField(fileScopedIdentifier),
+                  EntityTypeProtoEnum.FILES));
+            }
+          } catch (Exception e) {
+            log.error("Not able to add file reference");
```
Comment (mohitg0795) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java:100]: we should log the exception here.

**Comment by sourabh-awashti** [comment]
trigger uts

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java (hunk 5 above)
Comment (sourabh-awashti) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/CDPMSStepFilterJsonCreatorV2.java]: done

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration-steps/src/test/resources/schema/ApprovalStage/all.json` line 6379 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZKVSpCOLLrv1Y23l_:-NZZKVSpCOLLrv1Y23la:bllhj51) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/pipeline-service/modules/orchestration-steps/src/test/resources/schema/ApprovalStage/all.json#L6379)):*
> ```JSON
>       ],
>       "properties": {
>         "fileRefs": {
> ```

We will need to update static schema too for this.

**Review by sahilhindwani-harness — DISMISSED** [review]
> Reviewed 2 of 5 files at r1, 3 of 6 files at r5, 1 of 1 files at r6, 4 of 5 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r1, 3 of 5 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)* status: 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/test/java/io/harness/cdng/creator/plan/CDPMSStepFilterJsonCreatorV2Test.java` line 53 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZPsylDt56coC6qFM8:-NZZPsylDt56coC6qFM9:b-klx8ip) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/125-cd-nextgen/src/test/java/io/harness/cdng/creator/plan/CDPMSStepFilterJsonCreatorV2Test.java#L53)):*
> ```Java
> import org.mockito.MockitoAnnotations;
> 
> public class CDPMSStepFilterJsonCreatorV2Test extends CDNGTestBase {
> ```

you dont need this testBase, use categoryTest, example - DeploymentStageFilterJsonCreatorV2Test
___
*[`125-cd-nextgen/src/test/resources/schema/DeploymentStage/all.json` line 14913 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZQ6KO5KfUmmf-Bu8n:-NZZQ6KO5KfUmmf-Bu8o:b-6x09m6) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/125-cd-nextgen/src/test/resources/schema/DeploymentStage/all.json#L14913)):*
> ```JSON
>             "fileRef": {
>               "type": "string"
>             },
> ```

is this expected? @yogesh-chauhan 
___
*[`880-pipeline-cd-commons/src/main/java/io/harness/steps/shellscript/ShellScriptSourceWrapper.java` line 53 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZQ3ThAxoqBtVu5jFP:-NZZQ3ThAxoqBtVu5jFQ:b-51nfo2) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/880-pipeline-cd-commons/src/main/java/io/harness/steps/shellscript/ShellScriptSourceWrapper.java#L53)):*
> ```Java
>   }
> 
>   public List<String> getFileRefs() {
> ```

We should add @jsonIgnore, else they will come in schema

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)* status: 1 change requests. 1. unresolved discussions.

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/test/resources/schema/DeploymentStage/all.json` line 14913 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZQ6KO5KfUmmf-Bu8n:-NZZUtlmD6EGwnR_dtWS:b-1f9lfn) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/125-cd-nextgen/src/test/resources/schema/DeploymentStage/all.json#L14913)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

is this expected? @yogesh-chauhan 
</blockquote></details>

should not be. @sourabh-awashti this is due to the getFileRef method. Lets rename it

**Comment by sourabh-awashti** [comment]
*[`125-cd-nextgen/src/test/java/io/harness/cdng/creator/plan/CDPMSStepFilterJsonCreatorV2Test.java` line 53 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZPsylDt56coC6qFM8:-NZZhYUfAUaVTzMM5G4s:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/125-cd-nextgen/src/test/java/io/harness/cdng/creator/plan/CDPMSStepFilterJsonCreatorV2Test.java#L53)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

you dont need this testBase, use categoryTest, example - DeploymentStageFilterJsonCreatorV2Test
</blockquote></details>

Done.

**Comment by sourabh-awashti** [comment]
*[`880-pipeline-cd-commons/src/main/java/io/harness/steps/shellscript/ShellScriptSourceWrapper.java` line 53 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZQ3ThAxoqBtVu5jFP:-NZZhZzSA0d-8WA-vLIW:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/880-pipeline-cd-commons/src/main/java/io/harness/steps/shellscript/ShellScriptSourceWrapper.java#L53)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

We should add @jsonIgnore, else they will come in schema
</blockquote></details>

Done.

**Comment by sourabh-awashti** [comment]
*[`125-cd-nextgen/src/test/resources/schema/DeploymentStage/all.json` line 14913 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZQ6KO5KfUmmf-Bu8n:-NZZhaAu4z7BkIIqF_9r:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/125-cd-nextgen/src/test/resources/schema/DeploymentStage/all.json#L14913)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

should not be. @sourabh-awashti this is due to the getFileRef method. Lets rename it 
</blockquote></details>

Done.

**Comment by sourabh-awashti** [comment]
*[`pipeline-service/modules/orchestration-steps/src/test/resources/schema/ApprovalStage/all.json` line 6379 at r8](https://reviewable.io/reviews/harness/harness-core/49938#-NZZKVSpCOLLrv1Y23l_:-NZZhbg0APo_Hys0yYhb:ba32pwy) ([raw file](https://github.com/harness/harness-core/blob/238c1b2baa1575ecd5b1763bbfd07f02439ba348/pipeline-service/modules/orchestration-steps/src/test/resources/schema/ApprovalStage/all.json#L6379)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

We will need to update static schema too for this.
</blockquote></details>

no need to do as there is no schema change now

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger uts

**Review by archit-harness — APPROVED** [review]
> Reviewed 8 of 9 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)* status: 9 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`880-pipeline-cd-commons/src/main/java/io/harness/steps/shellscript/ShellScriptSourceWrapper.java` line 54 at r10](https://reviewable.io/reviews/harness/harness-core/49938#-NZ_N-n99cfHKUmt1qOK:-NZ_N-n99cfHKUmt1qOL:b74q46b) ([raw file](https://github.com/harness/harness-core/blob/4d22becd112de760100b5bba060d7baf6c6eafa2/880-pipeline-cd-commons/src/main/java/io/harness/steps/shellscript/ShellScriptSourceWrapper.java#L54)):*
> ```Java
>   }
> 
>   @JsonIgnore
> ```

we ideally don't need this now



_Code quote:_
```Java
@JsonIgnore
```

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 8 of 9 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49938)* status: 2. unresolved discussions.

**Comment by yogesh-chauhan** [comment]
trigger review

**Pull request merged**